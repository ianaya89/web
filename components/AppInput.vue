<template>
  <div>
    <span v-if="soloLabel && outline" class="text-grey text-p3 uppercase tracking-wide">
      {{ soloLabel }}
    </span>
    <span v-if="soloLabel && required && outline" class="text-red text-p3">
      *
    </span>
    <app-tooltip v-if="soloLabel && tooltip" class="text-p2">
      {{ tooltipText }}
    </app-tooltip>
    <v-text-field
      ref="textfield"
      v-bind="$props"
      :value="formattedValue"
      :class="[requiredClass, classes, labelMargins]"
      :type="inputType"
      :single-line="stickyLabel"
      :append-icon="displayedAppendIcon"
      :prepend-icon="displayedPrependIcon"
      :error-messages="translatableErrorMessages"
      @change="onChange"
      @input="onInput"
      @focus="onFocus"
      @blur="onBlur"
      @click:append="appendClick"
      @click:prepend="prependClick"
    />
  </div>
  <!-- if the label does not move up, maybe better name would be - staticLabel -->
</template>


<script>
import BaseInput from '@/components/BaseInput'
import AppTooltip from '@/components/AppTooltip'

export default BaseInput.extend({
  name: 'AppInput',

  components: {
    AppTooltip
  },

  props: {
    classes: {
      type: String,
      default: 'mt-6'
    },
    browserAutocomplete: {
      type: String,
      default: undefined
    },
    autofocus: {
      type: Boolean,
      default: false
    },
    type: {
      type: String,
      default: 'text',
      validator: value => {
        return ['text', 'email', 'password', 'number'].indexOf(value) !== -1
      }
    },
    isCurrency: {
      // supported only if type === 'text'
      type: Boolean,
      default: false
    },
    isArray: {
      // supported only if value.isArray()' useful for datepicker with multiple option
      type: Boolean,
      default: false
    },
    passwordSwitch: {
      type: Boolean,
      default: false
    },
    suffix: {
      type: String,
      default: null
    },
    counter: {
      type: Number,
      default: null
    },
    appendedIcon2: {
      type: String,
      default: null
    },
    prependedIcon2: {
      type: String,
      default: null
    },
    value: {
      type: [String, Number, Array],
      default: null
    },
    soloLabel: {
      type: String,
      default: 'Label'
    },
    tooltip: {
      type: Boolean,
      default: null
    },
    tooltipText: {
      type: String,
      default: 'Tooltip'
    },
    errorMessages: {
      type: Array,
      default() {
        return []
      }
    }
  },

  data: () => ({
    appendIconSwitch: false,
    prependIconSwitch: false,
    valueData: null
  }),

  computed: {
    translatableErrorMessages() {
      return this.errorMessages.map(message => this.$t(message))
    },

    hasError() {
      if (!this.$refs.textfield) {
        return false
      }

      return this.$refs.textfield.hasError
    },

    displayedAppendIcon() {
      if (this.appendedIcon2 == null) {
        return this.appendedIcon
      } else if (this.appendIconSwitch) {
        return this.appendedIcon2
      } else {
        return this.appendedIcon
      }
    },

    displayedPrependIcon() {
      if (this.prependedIcon2 == null) {
        return this.prependedIcon
      } else if (this.prependIconSwitch) {
        return this.prependedIcon2
      } else {
        return this.prependedIcon
      }
    },

    inputType() {
      if (this.passwordSwitch && this.appendIconSwitch) {
        return 'text'
      } else if (this.passwordSwitch && !this.appendIconSwitch) {
        return 'password'
      } else {
        return this.type
      }
    },

    requiredClass() {
      return this.required ? 'required-sign' : 'no-required'
    },

    labelMargins() {
      return this.label === null ? 'no-label' : ''
    },

    formattedValue() {
      if (this.isCurrency) {
        return this.$filters.currency(this.valueData, false, false, 0)
      } else if (this.isArray) {
        return this.valueData !== null ? this.valueData.join(', ') : ''
      } else {
        return this.valueData
      }
    }
  },

  watch: {
    value(newVal) {
      this.valueData = newVal
    }
  },

  mounted() {
    this.valueData = this.value
  },

  methods: {
    formatCurrencyInput(value) {
      if (this.isCurrency) {
        const input = this.$el.querySelector('input')
        const initPosition = input.selectionStart
        value = Number(
          String(value)
            .replace(/ /g, '')
            .replace(/,/g, '')
        )
        this.valueData = value
        this.$nextTick(() => {
          this.setCaretPosition(input, initPosition)
        })
      }
      return value
    },

    appendClick() {
      this.appendIconSwitch = !this.appendIconSwitch
    },

    prependClick() {
      this.prependIconSwitch = !this.prependIconSwitch
    },

    focus() {
      this.$refs.textfield.focus()
    },

    onFocus() {
      this.$emit('focus')
    },

    onBlur() {
      this.$emit('blur')
    },

    onChange(value) {
      value = this.formatCurrencyInput(value)
      if (this.type === 'number') {
        this.$emit('change', Number(value))
      } else {
        this.$emit('change', value)
      }
    },

    onInput(value) {
      value = this.formatCurrencyInput(value)
      if (this.type === 'number') {
        this.$emit('input', Number(value))
      } else {
        this.$emit('input', value)
      }
    },

    setCaretPosition(elem, caretPos) {
      if (elem != null) {
        elem.focus()
        elem.setSelectionRange(caretPos, caretPos)
      }
    },

    resetValidation() {
      if (this.$refs.textfield) {
        this.$refs.textfield.resetValidation()
      }
    },

    validate() {
      if (this.$refs.textfield) {
        this.$refs.textfield.validate(true)
      }
    }
  }
})
</script>

<style lang="postcss">
.v-text-field {
  input {
    @apply py-0;
  }
}

.v-input {
  input {
    max-height: none;

    &[type='password'] {
      @apply .font-serif;
    }
  }
}

.v-input.v-text-field {
  label {
    @apply .tracking-wide .uppercase;
    top: 0.15rem;
  }

  &.no-label input {
    @apply .my-auto;
  }

  .v-icon {
    @apply .text-menu .text-red;
    &.text-red {
      @apply .text-red;
    }
  }

  &.before-after-white {
    .v-input__control .v-input__slot {
      &:hover {
        &:before {
          @apply .border-white;
        }
      }

      &:before {
        border-color: rgba(255, 255, 255, 0.6);
      }
    }
  }

  .v-messages {
    @apply .pt-1 .leading-tight .normal-case;
  }

  .v-counter {
    @apply .pt-2;
  }

  .v-icon {
    @apply .text-menu;
  }

  .v-input__slot {
    margin-bottom: 0;
  }

  &.text-right .v-text-field__slot input {
    @apply .text-right;
  }

  &.text-h4 .v-text-field__slot input {
    @apply .text-h4;
  }

  &.text-white .v-text-field__slot input {
    @apply .text-white;
  }

  &.v-text-field .v-label {
    @apply .tracking-wide .truncate;
    word-wrap: break-word;
  }
}

.v-text-field--outline {
  .v-input__control {
    min-height: 5.5rem;
    .v-input__slot {
      @apply .rounded-none .border .border-grey-l2 !important;
      transition: 0.15s ease-in-out !important;

      &:hover,
      &:active {
        /* @apply .border-black

        // !important; */
      }
    }
  }
}

.v-text-field--outline.v-input--has-state {
  .v-input__control > .v-input__slot {
    border: 1px solid red !important;
  }
}

.v-text-field.v-input--has-state {
  .v-input__control > .v-input__slot::before {
    @apply .border-red;
  }

  .v-messages__message {
    @apply .text-red .leading-tight;
  }
}

.v-input__control {
  @apply .tracking-wide;

  input,
  textarea {
    @apply .text-menu .font-medium;
  }
}

.v-input--is-disabled {
  opacity: 0.6;
}

.required-sign {
  label::after {
    @apply .text-red;
    content: '*';
  }
}

.no-required {
}

.v-text-field--outline > .v-input__control > .v-input__slot {
  background: white !important;
}
</style>
