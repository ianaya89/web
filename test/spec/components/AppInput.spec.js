import { shallowMount, mount } from '@vue/test-utils'
import AppInput from '@/components/AppInput'

describe('AppInput', () => {
  it('renders default input', () => {
    const wrapper = shallowMount(AppInput, {
      propsData: {
        appendIconClickCallback: jest.fn(),
        prependIconClickCallback: jest.fn()
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders email sticky label input', () => {
    const wrapper = shallowMount(AppInput, {
      propsData: {
        type: 'email',
        stickyLabel: true,
        label: 'Email',
        appendIconClickCallback: jest.fn(),
        prependIconClickCallback: jest.fn()
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders input with placeholder outline and hint', () => {
    const wrapper = shallowMount(AppInput, {
      propsData: {
        placeholder: 'Your name',
        outline: true,
        label: 'First name',
        hint: 'Please write your first name',
        persistentHint: true,
        appendIconClickCallback: jest.fn(),
        prependIconClickCallback: jest.fn()
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('check input value', async () => {
    const wrapper = mount(AppInput, {
      propsData: {
        value: 'Hello',
        appendIconClickCallback: jest.fn(),
        prependIconClickCallback: jest.fn()
      }
    })
    await wrapper.vm.$nextTick()
    const input = wrapper.find('input')
    expect(input.element.value).toBe('Hello')
    await input.setValue('value')
    expect(wrapper.emitted('input')).toBeTruthy()
    expect(wrapper.emitted('input')[0][0]).toEqual('value')
  })

  it('check password visible option', () => {
    const wrapper = mount(AppInput, {
      propsData: {
        type: 'password',
        passwordSwitch: true,
        appendedIcon: 'lni-eye',
        appendedIcon2: 'lni-cross-circle',
        appendIconClickCallback: jest.fn(),
        prependIconClickCallback: jest.fn()
      }
    })
    const appendIcon = wrapper.find('.v-input__icon--append .v-icon')

    expect(wrapper.element).toMatchSnapshot()
    appendIcon.trigger('click')
    expect(wrapper.element).toMatchSnapshot()
    appendIcon.trigger('click')
    expect(wrapper.element).toMatchSnapshot()
  })

  it('check formatting currency value not affect event value', () => {
    const wrapper = mount(AppInput, {
      propsData: {
        isCurrency: true,
        appendIconClickCallback: jest.fn(),
        prependIconClickCallback: jest.fn()
      }
    })

    const input = wrapper.find('input')
    input.setValue(123123)
    expect(wrapper.emitted('input')).toBeTruthy()
    expect(wrapper.emitted('input')[0][0]).toEqual(123123)
  })

  it('triggers input event with correct value format after input changes', () => {
    let wrapper = mount(AppInput, {
      propsData: {
        type: 'number'
      }
    })

    let input = wrapper.find('input')
    input.setValue('1234')

    expect(wrapper.emitted('input')).toBeTruthy()
    expect(wrapper.emitted('input')[0][0]).toEqual(Number(1234))

    wrapper = mount(AppInput)

    input = wrapper.find('input')
    input.setValue('1234')

    expect(wrapper.emitted('input')).toBeTruthy()
    expect(wrapper.emitted('input')[0][0]).toEqual('1234')
  })

  it('triggers change event with correct value format after onChange method', () => {
    let wrapper = mount(AppInput, {
      propsData: {
        type: 'number'
      }
    })

    wrapper.vm.onChange('1234')

    expect(wrapper.emitted('change')).toBeTruthy()
    expect(wrapper.emitted('change')[0][0]).toEqual(Number(1234))

    wrapper = mount(AppInput)

    wrapper.vm.onChange('1234')

    expect(wrapper.emitted('change')).toBeTruthy()
    expect(wrapper.emitted('change')[0][0]).toEqual('1234')
  })

  it('validate input field', () => {
    const wrapper = mount(AppInput, {
      propsData: {
        required: true,
        rules: [v => !!v || 'required']
      }
    })

    wrapper.vm.validate()

    expect(wrapper.vm.$refs.textfield.valid).toBeFalsy()

    const input = wrapper.find('input')
    input.setValue(123123)

    wrapper.vm.validate()

    expect(wrapper.vm.$refs.textfield.valid).toBeTruthy()
  })

  it('reset validation', () => {
    const wrapper = mount(AppInput, {
      propsData: {
        required: true,
        rules: [v => !!v || 'required']
      }
    })

    wrapper.vm.validate()

    expect(wrapper.vm.$refs.textfield.hasMessages).toBeTruthy()

    wrapper.vm.resetValidation()

    expect(wrapper.vm.$refs.textfield.hasMessages).toBeFalsy()
  })
})
