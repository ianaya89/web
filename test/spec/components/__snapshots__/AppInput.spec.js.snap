// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInput check password visible option 1`] = `
<div
  appendiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
  prependiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="v-input v-text-field theme--light no-required mt-6"
  >
    <div
      class="v-input__control"
    >
      <div
        class="v-input__slot"
      >
        <div
          class="v-text-field__slot"
        >
          <label
            aria-hidden="true"
            class="v-label theme--light"
            style="left: 0px; position: absolute;"
          >
            Label
          </label>
          <input
            appendedicon="lni-eye"
            appendedicon2="lni-cross-circle"
            aria-label="Label"
            classes="mt-6"
            passwordswitch="true"
            sololabel="Label"
            tooltiptext="Tooltip"
            type="password"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon v-icon--link lni lni-eye theme--light"
            />
          </div>
        </div>
      </div>
      <div
        class="v-text-field__details"
      >
        <div
          class="v-messages theme--light"
        >
          <div
            class="v-messages__wrapper"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppInput check password visible option 2`] = `
<div
  appendiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
  prependiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="v-input v-text-field theme--light no-required mt-6"
  >
    <div
      class="v-input__control"
    >
      <div
        class="v-input__slot"
      >
        <div
          class="v-text-field__slot"
        >
          <label
            aria-hidden="true"
            class="v-label theme--light"
            style="left: 0px; position: absolute;"
          >
            Label
          </label>
          <input
            appendedicon="lni-eye"
            appendedicon2="lni-cross-circle"
            aria-label="Label"
            classes="mt-6"
            passwordswitch="true"
            sololabel="Label"
            tooltiptext="Tooltip"
            type="password"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon v-icon--link lni lni-eye theme--light"
            />
          </div>
        </div>
      </div>
      <div
        class="v-text-field__details"
      >
        <div
          class="v-messages theme--light"
        >
          <div
            class="v-messages__wrapper"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppInput check password visible option 3`] = `
<div
  appendiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
  prependiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <div
    class="v-input v-text-field theme--light no-required mt-6"
  >
    <div
      class="v-input__control"
    >
      <div
        class="v-input__slot"
      >
        <div
          class="v-text-field__slot"
        >
          <label
            aria-hidden="true"
            class="v-label theme--light"
            style="left: 0px; position: absolute;"
          >
            Label
          </label>
          <input
            appendedicon="lni-eye"
            appendedicon2="lni-cross-circle"
            aria-label="Label"
            classes="mt-6"
            passwordswitch="true"
            sololabel="Label"
            tooltiptext="Tooltip"
            type="password"
          />
        </div>
        <div
          class="v-input__append-inner"
        >
          <div
            class="v-input__icon v-input__icon--append"
          >
            <i
              aria-hidden="true"
              class="v-icon v-icon--link lni lni-eye theme--light"
            />
          </div>
        </div>
      </div>
      <div
        class="v-text-field__details"
      >
        <div
          class="v-messages theme--light"
        >
          <div
            class="v-messages__wrapper"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppInput renders default input 1`] = `
<div
  appendiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
  prependiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <v-text-field-stub
    class=" 0 1"
    classes="mt-6"
    label="Label"
    rules=""
    sololabel="Label"
    tooltiptext="Tooltip"
    type="text"
  />
</div>
`;

exports[`AppInput renders email sticky label input 1`] = `
<div
  appendiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
  prependiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
>
  <!---->
   
  <!---->
   
  <!---->
   
  <v-text-field-stub
    class=" 0 1"
    classes="mt-6"
    label="Email"
    rules=""
    single-line="true"
    sololabel="Label"
    stickylabel="true"
    tooltiptext="Tooltip"
    type="email"
  />
</div>
`;

exports[`AppInput renders input with placeholder outline and hint 1`] = `
<div
  appendiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
  prependiconclickcallback="function mockConstructor() {
        return fn.apply(this, arguments);
      }"
>
  <span
    class="text-grey text-p3 uppercase tracking-wide"
  >
    
    Label
  
  </span>
   
  <!---->
   
  <!---->
   
  <v-text-field-stub
    class=" 0 1"
    classes="mt-6"
    hint="Please write your first name"
    label="First name"
    outline="true"
    persistenthint="true"
    placeholder="Your name"
    rules=""
    sololabel="Label"
    tooltiptext="Tooltip"
    type="text"
  />
</div>
`;
